src/StorageFolding.cpp

```diff
@@ -99,31 +109,71 @@ class AttemptStorageFoldingOfFunction : public IRMutator {
+                    min_monotonic_increasing = true;
+                } else {
+                    Expr max_next = substitute(op->name, loop_var + 1, max);
+                    condition = max_next <= max;
+
+                    // After we assert that the max decreased, assume
+                    // the max is monotonically decreasing.
+                    max_monotonic_decreasing = true;
+                }
+                Expr error = Call::make(Int(32), "halide_error_bad_fold",
+                                        {func.name(), storage_dim.var, op->name},
+                                        Call::Extern);
+
+                body = Block::make(AssertStmt::make(condition, error), body);
+            }
+
             // The min or max has to be monotonic with the loop
             // variable, and should depend on the loop variable.
-            if (min_monotonic_increasing ||
-                max_monotonic_decreasing) {
-
-                // The max of the extent over all values of the loop variable must be a constant
+            if (min_monotonic_increasing || max_monotonic_decreasing) {
                 Expr extent = simplify(max - min);
-                Scope<Interval> scope;
-                scope.push(op->name, Interval(Variable::make(Int(32), op->name + ".loop_min"),
-                                              Variable::make(Int(32), op->name + ".loop_max")));
-                Expr max_extent = bounds_of_expr_in_scope(extent, scope).max;
-                scope.pop(op->name);
-
-                max_extent = find_constant_bound(simplify(max_extent), Direction::Upper);
+                Expr factor;
+                if (explicit_factor.defined()) {
+                    Expr error = Call::make(Int(32), "halide_error_fold_factor_too_small",
+                                            {func.name(), storage_dim.var, explicit_factor, op->name, extent},
+                                            Call::Extern);
+                    body = Block::make(AssertStmt::make(extent <= explicit_factor, error), body);
+
+                    factor = explicit_factor;
+                } else {
+                    // The max of the extent over all values of the loop variable must be a constant
+                    Scope<Interval> scope;
+                    scope.push(op->name, Interval(Variable::make(Int(32), op->name + ".loop_min"),
+                                                  Variable::make(Int(32), op->name + ".loop_max")));
+                    Expr max_extent = simplify(bounds_of_expr_in_scope(extent, scope).max);
+                    scope.pop(op->name);
 
-                if (const int64_t *extent = as_const_int(max_extent)) {
+                    max_extent = find_constant_bound(max_extent, Direction::Upper);
 
-                    int factor = 1;
-                    while (factor <= *extent && factor < 1024) factor *= 2;
+                    if (const int64_t *extent = as_const_int(max_extent)) {
```

Comment:
I did intentionally remove it, because I think it can only result in incorrect code. If it hits that limit, then the generated code will fold into a smaller buffer than it should, which seems bad. I think the 1024 limit was just there as a "reasonable upper bound that will never be hit". @abadams can you confirm? Should the 1024 limit be removed?